.circleci/config.yml

```diff
@@ -265,67 +402,38 @@ jobs:
           paths:
             - ui
 
-  # Run suite of Puppeteer end-to-end QA tests, pointed at the Workbench "test" environment.
+  # Run Puppeteer end-to-end UI tests on Workbench deployed "test" environment.
   puppeteer-e2e-test:
     <<: *defaults
     docker:
-      - image: circleci/node:13.10.1-browsers
-        environment:
-          NODE_ENV: development
+      - image: circleci/node:12.16.2-browsers
+    environment:
+      WORKBENCH_ENV: dev
+      CI: true
+      NODE_ENV: development
     working_directory: ~/workbench
     parallelism: 3
     steps:
-      - checkout
-      - run: *update_git_submodules
-      - restore_cache:
-          keys:
-          - v2-e2e-test-cache-{{ checksum "~/workbench/e2e/yarn.lock" }}
-          - v2-e2e-test-cache-
-      - run:
-          name: Initialize Yarn
-          working_directory: ~/workbench/e2e
-          command: yarn cache clean && yarn install --frozen-lockfile
-      - save_cache:
-          paths:
-            - ~/.cache/yarn
-          key: v2-e2e-test-cache-{{ checksum "~/workbench/e2e/yarn.lock" }}
-      - run:
-          name: Update Puppeteer Environment Variables
-          command: |
-            echo 'export USER_NAME=$PUPPETEER_TEST_USER' >> $BASH_ENV
-            echo 'export PASSWORD=$PUPPETEER_TEST_USER_PASSWORD' >> $BASH_ENV
-            echo 'export INVITATION_KEY=$PUPPETEER_TEST_REGISTRATION_KEY' >> $BASH_ENV
-            source $BASH_ENV
-      - run:
-          name: Run e2e tests (Puppeteer and Jest)
-          working_directory: ~/workbench/e2e
-          command: yarn test:ci $(circleci tests glob "tests/**/*.spec.ts" | circleci tests split --split-by=timings --timings-type=classname --show-counts)
-      - store_artifacts:
-          path: e2e/logs
-          destination: logs
-      - store_test_results:
-          path: ~/workbench/e2e/logs
-      
+      - git_checkout
+      - e2e_dependencies
+      - run_e2e_test
+
   ui-deploy-to-test:
     <<: *defaults
     steps:
-      - checkout
-      - run: *update_git_submodules
-      # Use the ui-build-test workspace here to avoid redoing the setup.
+      - git_checkout
+      # Use "ui-build-test" job's workspace here to avoid redoing the setup.
```

Comment:
Can we not do this in an idempotent way with the cache facility? I'm nervous that we've introduced an order dependency here.